src/Cache/FileHandler.php

```diff
@@ -41,21 +41,65 @@ public function read(): ?CacheInterface
             return null;
         }
 
-        $content = file_get_contents($this->file);
-
-        try {
-            $cache = Cache::fromJson($content);
-        } catch (\InvalidArgumentException $exception) {
+        $handle = fopen($this->file, 'r');
+        if (false === $handle) {
             return null;
         }
 
+        $cache = $this->readFromHandle($handle);
+
+        fclose($handle);
+
         return $cache;
     }
 
     public function write(CacheInterface $cache): void
     {
-        $content = $cache->toJson();
+        $this->ensureFileIsWriteable();
+
+        $handle = fopen($this->file, 'r+');
+        if (false === $handle) {
+            return;
+        }
+
+        if (getenv('PHP_CS_FIXER_EXPERIMENTAL_PARALLEL_CACHE')) {
```

Comment:
idea for future (out of this PR scope) - we can detect that file was modified since it was read initially, and do the backilling based on that indicator, without env var.

Reply:
implemented in 2aef0f81b5588e19b6ed9ac9d03b885614a23a15

```
ker@d ~/github/PHP-CS-Fixer λ ./php-cs-fixer list-files --config=.php-cs-fixer.dist.php | xargs -n 400 -P 3 ./php-cs-fixer fix --config=.php-cs-fixer.dist.php --path-mode intersection   
Loaded config default from ".php-cs-fixer.dist.php".
Using cache file ".php-cs-fixer.cache".
Loaded config default from ".php-cs-fixer.dist.php".
Using cache file ".php-cs-fixer.cache".
Loaded config default from ".php-cs-fixer.dist.php".
Using cache file ".php-cs-fixer.cache".
string(34) "
73993 !!! FRS !!! READ 1689350710"
string(34) "
73994 !!! FRS !!! READ 1689350710"
string(34) "
73995 !!! FRS !!! READ 1689350710"

Fixed 0 of 212 files in 7.380 seconds, 20.000 MB memory used
string(31) "
73995 !!! FRS !!! DECISION ???"
array(3) {
  ["property"]=>
  int(1689350710)
  ["os"]=>
  int(1689350710)
  ["should backfill"]=>
  bool(false)
}
string(28) "
73995 !!! FRS !!! SAVE FILE"
string(40) "
73995 !!! FRS !!! AFTER SAVE 1689350727"

Fixed 0 of 400 files in 23.354 seconds, 22.000 MB memory used
string(31) "
73993 !!! FRS !!! DECISION ???"
array(3) {
  ["property"]=>
  int(1689350710)
  ["os"]=>
  int(1689350727)
  ["should backfill"]=>
  bool(true)
}
string(27) "
73993 !!! FRS !!! BACKFILL"
string(28) "
73993 !!! FRS !!! SAVE FILE"
string(40) "
73993 !!! FRS !!! AFTER SAVE 1689350743"
string(31) "
73994 !!! FRS !!! DECISION ???"
array(3) {
  ["property"]=>
  int(1689350710)
  ["os"]=>
  int(1689350743)
  ["should backfill"]=>
  bool(true)
}
string(27) "
73994 !!! FRS !!! BACKFILL"
string(28) "
73994 !!! FRS !!! SAVE FILE"
string(40) "
73994 !!! FRS !!! AFTER SAVE 1689350748"

Fixed 0 of 400 files in 27.835 seconds, 22.000 MB memory used
```